plutus-ir/src/Language/PlutusIR/Analysis/Dependencies.hs

```diff
@@ -0,0 +1,127 @@
+module Language.PlutusIR.Analysis.Dependencies (Node (..), DepGraph, runTermDeps, runTypeDeps) where
+
+import qualified Language.PlutusCore               as PLC
+import qualified Language.PlutusCore.Name          as PLC
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Analysis.Usages as Usages
+
+import           Control.Monad.Reader
+
+import qualified Algebra.Graph.Class               as G
+import           Data.Coerce
+import qualified Data.Set                          as Set
+import           Lens.Micro
+
+-- | A node in a dependency graph. Either a specific 'PLC.Unique', or a specific
+-- node indicating the root of the graph. We need the root node because when computing the
+-- dependency graph of, say, a term, there will not be a binding for the term itself which
+-- we can use to represent it in the graph.
+data Node = Unique PLC.Unique | Root deriving (Show, Eq, Ord)
+
+-- | A constraint requiring @g@ to be a 'G.Graph' (so we can compute e.g. a @Relation@ from it), whose
+-- vertices are 'Node's.
+type DepGraph g = (G.Graph g, (G.Vertex g)~Node)
+
+-- | Compute the dependency graph of a 'Term'. The 'Root' node will correspond to the term itself.
+runTermDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+    => Term tyname name a
+    -> g
+runTermDeps t = runReader (termDeps t) Root
+
+-- | Compute the dependency graph of a 'Type'. The 'Root' node will correspond to the type itself.
+runTypeDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Type tyname a
+    -> g
+runTypeDeps t = runReader (typeDeps t) Root
+
+-- | Record some dependencies on the current node.
+recordDeps
+    :: (DepGraph g, MonadReader Node m)
+    => [PLC.Unique]
+    -> m g
+recordDeps us = do
+    current <- ask
+    pure $ G.connect (G.vertices [current]) (G.vertices (fmap Unique us))
+
+-- | Process the given action with the given name as the current node.
+withCurrent
+    :: (MonadReader Node m, PLC.HasUnique n u)
+    => n
+    -> m g
+    -> m g
+withCurrent n = local (const $ Unique $ coerce $ n ^. PLC.unique)
+
+bindingDeps
+    :: (DepGraph g, MonadReader Node m, PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+    => Binding tyname name a
+    -> m g
+bindingDeps b = case b of
+    TermBind _ d@(VarDecl _ n _) rhs -> do
+        vDeps <- varDeclDeps d
+        tDeps <- withCurrent n $ termDeps rhs
+        pure $ G.overlay vDeps tDeps
+    TypeBind _ d@(TyVarDecl _ n _) rhs -> do
+        vDeps <- tyVarDeclDeps d
+        tDeps <- withCurrent n $ typeDeps rhs
+        pure $ G.overlay vDeps tDeps
+    DatatypeBind _ (Datatype _ d tvs _ constrs) -> do
+        vDeps <- tyVarDeclDeps d
+        tvDeps <- traverse tyVarDeclDeps tvs
+        cstrDeps <- traverse varDeclDeps constrs
+        pure $ G.overlays $ [vDeps] ++ tvDeps ++ cstrDeps
```

Comment:
Looking at this very nice applicativey code, I can't not say it can be parallelized.

Reply:
You are probably right. I rewrote this all to just require an applicative, but I'm not sure it's any clearer...

Reply:
Oh, I didn't propose to do that, sorry. I meant, if we realize that speeding this up results in improved performance, then we can parallelize that. The style is fine either way for me.

Reply:
Yeah, it's *nearly* nice. I'm going to back out the change, but it's definitely doable.